File: Bronze1-1/06-21-pm/unit04/largest-divisible-subsequence/divsubseq.java
/*
 * Longest divisible subsequence
 */
import java.util.*;
public class divsubseq {
	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		int N = in.nextInt(); //# of integers
		int[] A = new int[N];
		for(int j=0; j<N; j++)
			A[j] = in.nextInt();
		in.close();
		
		ArrayList<ArrayList<Integer>> lists = new ArrayList<>();
		for(int j=0; j<N; j++) {
			lists.add(new ArrayList<>());
			lists.get(j).add(A[j]);
		}
		
		for(int i=0; i<N; i++) {
			for(int j=i+1; j<N; j++) {
				if(A[j]%A[i]==0 && lists.get(j).size() <= lists.get(i).size()) {
					lists.get(j).clear();
					lists.get(j).addAll(lists.get(i));
					lists.get(j).add(A[j]);
				}
			}
		}

		int maxL = 1;
		int id = 0;
		for(int j=0; j<N; j++) {
			if(lists.get(j).size()>maxL) {
				maxL = lists.get(j).size();
				id = j;
			}
		}
		
		System.out.println(maxL);
		for(int v : lists.get(id))
			System.out.print(v + " ");
		System.out.println();
	}
}

File: Bronze1-1/06-21-pm/unit04/largest-divisible-subsequence/divsubseq.cpp
/*
 * Longest divisible subsequence
 */
#include <bits/stdc++.h>
using namespace std;

int main() {
	int N; cin >> N; //# of integers
	int A[N];
	for(int j=0; j<N; j++)
		cin >> A[j];
	
	vector<vector<int>> lists(N);
	for(int j=0; j<N; j++) {
		lists[j].push_back(A[j]);
	}
	
	for(int i=0; i<N; i++) {
		for(int j=i+1; j<N; j++) {
			if(A[j]%A[i]==0 && lists[j].size() <= lists[i].size()) {
				lists[j].clear();
				lists[j] = lists[i];
				lists[j].push_back(A[j]);
			}
		}
	}

	int maxL = 1;
	int id = 0;
	for(int j=0; j<N; j++) {
		if(lists[j].size()>maxL) {
			maxL = lists[j].size();
			id = j;
		}
	}
	
	cout << maxL << '\n';
	for(int v : lists[id])
		cout << v << " ";
	cout << '\n';
}

